File: core/policies.py
from statistics import mean, stdev
from typing import Any, Callable, Dict, List, Optional, Tuple

from .criterion import by_win_rate
from .text import jaccard_similarity, normalize_and_tokenize


def group_bills_by_structure(
    bills: List[Tuple[str, str, float]],
    threshold: float = 0.75,
) -> List[Dict[str, Any]]:
    """
    Agrupa projetos de lei por similaridade de estrutura (Jaccard).
    """
    if not bills:
        return []

    groups: List[Dict[str, Any]] = []

    for municipio, frase, score in bills:
        tokens = normalize_and_tokenize(frase)

        if not groups:
            groups.append({
                "rep_tokens": tokens,
                "rep_phrase": frase,
                "members": [(municipio, frase, score, 1.0)],
            })
            continue

        best_idx: Optional[int] = None
        best_sim = 0.0

        for i, g in enumerate(groups):
            sim = jaccard_similarity(tokens, g["rep_tokens"])
            if sim > best_sim:
                best_sim = sim
                best_idx = i

        if best_idx is not None and best_sim >= threshold:
            groups[best_idx]["members"].append((municipio, frase, score, best_sim))
        else:
            groups.append({
                "rep_tokens": tokens,
                "rep_phrase": frase,
                "members": [(municipio, frase, score, 1.0)],
            })

    return groups


def compute_mean_and_std(values: List[float]) -> Tuple[float, float]:
    if not values:
        return 0.0, 0.0
    if len(values) == 1:
        return values[0], 0.0
    return mean(values), stdev(values)


def generate_policies_from_bills(
    bills: List[Tuple[str, str, float]],
    min_group_members: int = 2,
    similarity_threshold: float = 0.75,
    criterion: Callable[[List[float]], float] = by_win_rate,
) -> List[Dict[str, Any]]:
    """
    Recebe (municipio, descricao_PL, efeito) e devolve candidatos a políticas.
    """
    groups = group_bills_by_structure(bills, threshold=similarity_threshold)
    policies: List[Dict[str, Any]] = []

    for g in groups:
        members = g["members"]

        if len(members) < min_group_members:
            continue

        scores = [s for (_, _, s, _) in members]

        effect_mean, effect_std = compute_mean_and_std(scores)
        quality_score = criterion(scores) if criterion else 0.0
        actions = [(mun, frase, score) for (mun, frase, score, _) in members]

        policies.append({
            "policy": g["rep_phrase"],
            "actions": actions,
            "effect_mean": effect_mean,
            "effect_std": effect_std,
            "quality_score": quality_score,
        })

    policies.sort(key=lambda p: p["quality_score"], reverse=True)
    return policies

File: src/lib/policies.ts
import { jaccardSimilarity, normalizeAndTokenize } from "./text";

export type GroupedPolicyAction = {
  municipio: string;
  acao: string;
  score: number;
  rawEffect?: number | null;
};

export type PolicyCandidate = {
  policy: string;
  actions: GroupedPolicyAction[];
  effectMean: number;
  effectStd: number;
  qualityScore: number;
};

type GroupMember = {
  municipio: string;
  phrase: string;
  score: number;
  similarity: number;
  rawEffect?: number | null;
};

const computeMeanAndStd = (values: number[]): { mean: number; std: number } => {
  if (!values.length) return { mean: 0, std: 0 };
  if (values.length === 1) return { mean: values[0], std: 0 };
  const mean = values.reduce((sum, value) => sum + value, 0) / values.length;
  const variance =
    values.reduce((sum, value) => sum + (value - mean) * (value - mean), 0) / (values.length - 1 || 1);
  const std = Math.sqrt(variance);
  return { mean, std };
};

const byWinRate = (scores: number[]): number => {
  if (!scores.length) return 0;
  const n = scores.length;
  const winRate = scores.filter((s) => s < 0).length / n;
  const evidence = n / (n + 1);
  return winRate * evidence;
};

const bestMembersPerCity = (members: GroupMember[]): GroupMember[] => {
  const best = new Map<string, GroupMember>();
  for (const member of members) {
    const current = best.get(member.municipio);
    if (!current || member.score < current.score) {
      best.set(member.municipio, member);
    }
  }
  return Array.from(best.values());
};

const groupBillsByStructure = (bills: GroupedPolicyAction[], threshold: number): GroupMember[][] => {
  if (!bills.length) return [];
  const groups: { repTokens: string[]; repPhrase: string; members: GroupMember[] }[] = [];

  for (const bill of bills) {
    const tokens = normalizeAndTokenize(bill.acao);
    if (!groups.length) {
      groups.push({
        repTokens: tokens,
        repPhrase: bill.acao,
        members: [{ municipio: bill.municipio, phrase: bill.acao, score: bill.score, similarity: 1, rawEffect: bill.rawEffect }],
      });
      continue;
    }

    let bestIndex = -1;
    let bestSim = 0;
    groups.forEach((group, index) => {
      const sim = jaccardSimilarity(tokens, group.repTokens);
      if (sim > bestSim) {
        bestSim = sim;
        bestIndex = index;
      }
    });

    if (bestIndex !== -1 && bestSim >= threshold) {
      groups[bestIndex].members.push({
        municipio: bill.municipio,
        phrase: bill.acao,
        score: bill.score,
        similarity: bestSim,
        rawEffect: bill.rawEffect,
      });
    } else {
      groups.push({
        repTokens: tokens,
        repPhrase: bill.acao,
        members: [{ municipio: bill.municipio, phrase: bill.acao, score: bill.score, similarity: 1, rawEffect: bill.rawEffect }],
      });
    }
  }

  return groups.map((g) => g.members);
};

export const generatePoliciesFromBills = (
  bills: GroupedPolicyAction[],
  minGroupMembers: number,
  similarityThreshold: number,
): PolicyCandidate[] => {
  const grouped = groupBillsByStructure(bills, similarityThreshold);
  const candidates: PolicyCandidate[] = [];

  grouped.forEach((members) => {
    const bestPerCity = bestMembersPerCity(members);
    if (bestPerCity.length < minGroupMembers) return;

    const scores = bestPerCity.map((item) => item.score);
    const effects = bestPerCity
      .map((item) => item.rawEffect)
      .filter((value): value is number => value != null && Number.isFinite(value));

    const { mean, std } = computeMeanAndStd(effects.length ? effects : scores);
    const qualityScore = byWinRate(scores);
    const actions = bestPerCity.map((item) => ({
      municipio: item.municipio,
      acao: item.phrase,
      score: item.score,
      rawEffect: item.rawEffect ?? null,
    }));

    candidates.push({
      policy: members[0]?.phrase ?? "Política sem título",
      actions,
      effectMean: mean,
      effectStd: std,
      qualityScore,
    });
  });

  candidates.sort((a, b) => b.qualityScore - a.qualityScore);
  return candidates;
};
